### ISSUE TITLE
objective rank:ndcg strange behaviour

### ISSUE THREAD

**Issue by pavellevap** [dmlc/xgboost#2768]
## Environment info
Operating System: Linux

Package used (python/R/jvm/C++): python

`xgboost` version used: 0.6

python version: Python 3.5.2+

## Steps to reproduce

```python
import xgboost as xgb
import numpy as np

n = 1

X = []
y = []
group_sizes = []

for i in range(n):
    X += [[1, 0], [0, 1]]
    y += [1, 2]
    group_sizes += [2]

dtrain = xgb.DMatrix(data=np.array(X), label=np.array(y))
dtrain.set_group(np.array(group_sizes))

model = xgb.train(params={'objective': 'rank:ndcg'}, dtrain=dtrain, num_boost_round=10)
model.predict(xgb.DMatrix(np.array([[1, 0], [0, 1]])))
```

## Issue
For n < 5 model.predict returns [0.5, 0.5]. So the model thinks both objects have the same relevance, despite the fact that there was n groups where second object had higher label.
I expect model.predict to return higher relevance for the second object for any n. And in fact model starts to behave like I expect if I choose n >= 5 or if I change objective to 'rank:pairwise'.

**Comment by mohit-shrma** [comment]
I have seen that python package doesn't work as expected for ranking tasks. However, xgboost from commandline works as expected and gives different results from the python package for same parameters especially for ranking tasks.

**Comment by alexeib** [comment]
i am seeing this issue with n > 5 for my data (e.g. n = ~20 with ~800 items per group). when i get to n > 30 i start seeing outputs that are not 0.5 anymore but only a few different numbers. for example:
print(model.predict(dtest)) ->
[ 0.97558612  0.97558612  0.97558612 ...,  0.97558612  0.97558612
  0.6440345 ]
 for rank:pairwise, this does not happen

**Comment by wby19912010** [comment]
Excuse me but what is the purpose of setting group???

**Comment by JoshuaC3** [comment]
There seems to be a bit more info in the [catboost docs](https://tech.yandex.com/catboost/doc/dg/references/ndcg-docpage/) about this.

My understanding is that we group similar items so that we do not get data leaks for almost idential items. See scikit-learns docs on [Grouped CVFolds](http://scikit-learn.org/stable/modules/cross_validation.html#group-k-fold).

@wby19912010 btw happy birthday?
